### PR TITLE
Optimize ToArray conversions

### PR DESCRIPTION
**Pull request by bahusoid**
Optimized `collection.Select().ToArray()` calls with the following [extension](https://github.com/bahusoid/nhibernate-core/blob/2e6b9b8b333c59a167aba2bff82ed359162d68ad/src/NHibernate/Util/EnumerableExtensions.cs#L42-L53)

### CODE REVIEW

**Comment by hazzik** [comment]
Hi, @bahusoid. Nice optimization. I actually planed this for a few thousand years (figurally) now. I made a synthetic [benchmark](https://gist.github.com/hazzik/13e6c5be40d3e8424136c0315b9c27fb) for this. 

A few take aways from the benchmark: 

- The boxing of `this` to `ICollection<T>` kills the intention. Better to use framework's `Array.ConvertAll` / `List<T>.ConvertAll`. The manual implementation is actually slower than both `Select(x=>...).ToArray()` and `Array.ConvertAll(...)`
- Method groups are slow.

**Comment by bahusoid** [comment]
> The manual implementation is actually slower than both

It seems that's true only for .NET Core. On full .NET framework `Select(x=>...).ToArray()` is slower

**Comment by hazzik** [comment]
I've updated benchmark. The boxing to `ICollection<>` is the culprit. I did not run on the Full .NET as I don't have access to a windows machine atm. You can run it for yourself. You'll need to add `[ClrJob]` attribute to the class.

**Comment by hazzik** [comment]
Ok, I've run on full .NET. The version with boxing is indeed faster than Select (about 25% faster), however, it is still 3 times slower than version without boxing.

File: src/NHibernate/SqlCommand/SqlInsertBuilder.cs [hunk 1]
```diff
@@ -221,7 +220,7 @@ public SqlCommandInfo ToSqlCommandInfo()
 
 		public SqlType[] GetParametersTypeArray()
 		{
-			return (new List<SqlType>(new SafetyEnumerable<SqlType>(columns.Values))).ToArray();
```
Comment (hazzik) [inline src/NHibernate/SqlCommand/SqlInsertBuilder.cs]: There is a subtle difference between `SafetyEnumerable` and `OfType`. `SafetyEnumerable` will not filter out nulls, while `OfType` will. Not sure how this is relevant here.

Comment (bahusoid) [reply src/NHibernate/SqlCommand/SqlInsertBuilder.cs]: Good catch.. Better to restore original logic

**Comment by hazzik** [comment]
I propose to rename these extensions to `ConvertAll` to be on par with the framework. What do you think?

**Comment by hazzik** [comment]
Or maybe not.

**Comment by bahusoid** [comment]
>I propose to rename these extensions to ConvertAll

I'm not sure it's a good fit here. `ConvertAll` name doesn't imply that returning type might be changed.

**Pull request merged**